// @vivliostyle/cli/dist/commands/preview.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const preview_1 = require("../preview");
const util_1 = require("../util");
const preview_parser_1 = require("./preview.parser");
try {
    const program = preview_parser_1.setupPreviewParserProgram();
    program.parse(process.argv);
    const options = program.opts();
    preview_1.preview({
        input: (_a = program.args) === null || _a === void 0 ? void 0 : _a[0],
        configPath: options.config,
        theme: options.theme,
        size: options.size,
        cropMarks: options.cropMarks,
        bleed: options.bleed,
        cropOffset: options.cropOffset,
        css: options.css,
        style: options.style,
        userStyle: options.userStyle,
        singleDoc: options.singleDoc,
        quick: options.quick,
        title: options.title,
        author: options.author,
        language: options.language,
        verbose: options.verbose,
        timeout: options.timeout,
        sandbox: options.sandbox,
        executableBrowser: options.executableBrowser,
        http: options.http,
        viewer: options.viewer,
        browser: options.browser,
        executableChromium: options.executableChromium, // TODO: Remove it in the next major version up
    }).catch(util_1.gracefulError);
}
catch (err) {
    if (err instanceof Error) {
        util_1.gracefulError(err);
    }
}
